flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/DeclarativeSlotManager.java

```diff
@@ -647,7 +647,12 @@ private ResourceCounter tryFulfillRequirementsWithPendingSlots(
                     pendingSlots = allocationResult.getNewAvailableResources();
                     if (!allocationResult.isSuccessfulAllocating()
                             && sendNotEnoughResourceNotifications) {
-                        LOG.warn("Could not fulfill resource requirements of job {}.", jobId);
+                        // TODO (review): free slots are logged as zero here, but non-zero in
+                        // JobMaster.slotPoolService
```

Comment:
Can we resolve this TODO? I don't really understand what the problem is.

Reply:
This log statement happens on the `ResourceManager`. The other log statement happens on the `JobMaster`. It could be the case that the RM has assigned all its slots to the JM (e.g. the RM has 0 free slots) but the JM hasn't assigned them yet.

Reply:
I see, I'll remove the TODO then.
Thanks for clarifying.